src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs

```diff
@@ -256,182 +253,46 @@ private unsafe IAsyncResult BeginWaitForConnection(AsyncCallback callback, Objec
-            }
-
-            IOCancellationHelper cancellationHelper = afsar.AsyncState as IOCancellationHelper;
-            if (cancellationHelper != null)
-            {
-                cancellationHelper.SetOperationCompleted();
-            }
-
-            // Obtain the WaitHandle, but don't use public property in case we
-            // delay initialize the manual reset event in the future.
-            WaitHandle wh = afsar._waitHandle;
-            if (wh != null)
-            {
-                // We must block to ensure that ConnectionIOCallback has completed,
-                // and we should close the WaitHandle in here.  AsyncFSCallback
-                // and the hand-ported imitation version in COMThreadPool.cpp 
-                // are the only places that set this event.
-                using (wh)
-                {
-                    wh.WaitOne();
-                    Debug.Assert(afsar._isComplete == true, "NamedPipeServerStream::EndWaitForConnection - AsyncFSCallback didn't set _isComplete to true!");
-                }
-            }
-
-            // We should have freed the overlapped and set it to null either in the Begin
-            // method (if ConnectNamedPipe completed synchronously) or in AsyncWaitForConnectionCallback.
-            // If it is not nulled out, we should not be past the above wait:
-            Debug.Assert(afsar._overlapped == null);
-
-            // Now check for any error during the read.
-            if (afsar._errorCode != 0)
-            {
-                if (afsar._errorCode == Interop.mincore.Errors.ERROR_OPERATION_ABORTED)
-                {
-                    if (cancellationHelper != null)
-                    {
-                        cancellationHelper.ThrowIOOperationAborted();
-                    }
+                    case Interop.mincore.Errors.ERROR_IO_PENDING:
+                        break;
+
+                    // If we are here then the pipe is already connected, or there was an error
+                    // so we should unpin and free the overlapped.
+                    case Interop.mincore.Errors.ERROR_PIPE_CONNECTED:
+                        // IOCompletitionCallback will not be called because we completed synchronously.
+                        completionSource.ReleaseResources();
+                        if (State == PipeState.Connected)
+                        {
+                            throw new InvalidOperationException(SR.InvalidOperation_PipeAlreadyConnected);
+                        }
+                        completionSource.SetCompletedSynchronously();
+
+                        // We return a cached task instead of TaskCompletionSource's Task allowing the GC to collect it.
+                        return Task.CompletedTask;
+
+                    default:
+                        completionSource.ReleaseResources();
+                        throw Win32Marshal.GetExceptionForWin32Error(errorCode);
```

Comment:
Hmm.  Seems like this (and some of the above exceptions) really should have been returned via Task.FromException, but I see in the old version that we were throwing there, too, so I guess we should leave it.